merkle/src/merkle.rs

```diff
@@ -248,6 +258,241 @@ impl<E: Element> Clone for MmapStore<E> {
     }
 }
 
+/// File-mapping version of `MmapStore` with the added `new_with_path` method
+/// that allows to set its path (otherwise a temporary file is used which is
+/// cleaned up after we drop this structure).
+#[derive(Debug)]
+pub struct DiskMmapStore<E: Element> {
+    // Implementing the `store` with `Arc`/`RwLock` to avoid adding lifetimes
+    // parameters to the struct (which might have larger repercussions in the
+    // definitions of the `MerkleTree` and its consumers. Also used for its
+    // coordination mechanisms, but it's not clearly defined if `MerkleTree`
+    // (and `Store`) should be thread-safe (so they might be removed later).
+    store: Arc<RwLock<Option<MmapMut>>>,
+
+    len: usize,
+    _e: PhantomData<E>,
+    file: File,
+    // We need to save the `File` in case we're creating a `tempfile()`
+    // otherwise it will get cleaned after we return from `new()`.
+
+    // We cache the `store.len()` call to avoid accessing the store when
+    // it's offloaded. Not to be confused with `len`, this saves the total
+    // size of the `store` and the other one keeps track of used `E` slots
+    // in the `DiskMmapStore`.
+    store_size: usize,
+
+    // We save the arguments of `new_with_path` to reconstruct it and reload
+    // the `MmapMut` after offload has been called.
+    path: String,
+    // FIXME: Using `String` instead of `&Path` to avoid using lifetimes
+    //  or errors like "the size for values of type `[u8]` cannot be known
+    //  at compilation time" (with `Option<Box<Path>>` or similar).
+    size: Option<usize>,
+    // FIXME: Needed to reconstruct the `new_with_path` call, especially
+    //  since we are adjusting the size there. Ideally this value should
+    //  be deduced from the MT settings and not stored here.
+}
+
+impl<E: Element> ops::Deref for DiskMmapStore<E> {
+    type Target = [E];
+
+    fn deref(&self) -> &Self::Target {
+        unimplemented!()
+    }
+}
+
+impl<E: Element> Store<E> for DiskMmapStore<E> {
+    #[allow(unsafe_code)]
+    fn new(size: usize) -> Self {
+        let byte_len = E::byte_len() * size;
+        let file: File = tempfile().expect("couldn't create temp file");
+        file.set_len(byte_len as u64)
+            .expect(&format!("couldn't set len of {}", byte_len));
+
+        let mmap = unsafe { MmapMut::map_mut(&file).expect("couldn't create map_mut") };
+        let mmap_size = mmap.len();
+        DiskMmapStore {
+            store: Arc::new(RwLock::new(Some(mmap))),
+            len: 0,
+            _e: Default::default(),
+            file,
+            store_size: mmap_size,
+            path: "".to_string(),
```

Comment:
why not store the path of the tempfile?

Reply:
I'm not sure how to obtain that, the tempfile in Unix is created with `O_TMPFILE` and I don't think that has a specific path associated to it (or if there is even an API in the Rust package to obtain it).

Reply:
I see, that's fine then.